commons-core/src/test/scala/com/avsystem/commons/serialization/json/JsonStringInputOutputTest.scala

```diff
@@ -0,0 +1,206 @@
+
+import scala.collection.mutable.ListBuffer
+import scala.util.Random
+
+class JsonStringInputOutputTest extends FunSuite with SerializationTestUtils with Matchers {
+
+  import JsonStringInput.read
+  import JsonStringOutput.write
+
+  test("raw json list test") {
+    val jsons = List("123", "null", "\"str\"", "4.5", "[1,2,3]", "{\"a\": 123, \"b\": 3.14}")
+    val sb = new JStringBuilder
+    val output = new JsonStringOutput(sb)
+    val lo = output.writeList()
+    jsons.foreach(json => lo.writeElement().writeRawJson(json))
+    lo.finish()
+    val jsonList = sb.toString
+
+    assert(jsonList == jsons.mkString("[", ",", "]"))
+
+    val input = new JsonStringInput(new JsonReader(jsonList))
+    val li = input.readList()
+    val resBuilder = new ListBuffer[String]
+    while (li.hasNext) {
+      resBuilder += li.nextElement().readRawJson()
+    }
+    assert(resBuilder.result() == jsons)
+  }
+
+  test("raw json object test") {
+    val jsons = IListMap("a" -> "123", "b" -> "null", "c" -> "\"str\"", "d" -> "4.5", "e" -> "[1,2,3]", "f" -> "{\"a\": 123, \"b\": 3.14}")
+    val sb = new JStringBuilder
+    val output = new JsonStringOutput(sb)
+    val oo = output.writeObject()
+    jsons.foreach { case (fn, fv) => oo.writeField(fn).writeRawJson(fv) }
+    oo.finish()
+    val jsonList = sb.toString
+
+    assert(jsonList == jsons.map({ case (k, v) => s""""$k":$v""" }).mkString("{", ",", "}"))
+
+    val input = new JsonStringInput(new JsonReader(jsonList))
+    val oi = input.readObject()
+    val resBuilder = IListMap.newBuilder[String, String]
+    while (oi.hasNext) {
+      val fi = oi.nextField()
+      resBuilder += ((fi.fieldName, fi.readRawJson()))
+    }
+    assert(resBuilder.result() == jsons)
+  }
+
+  def roundtrip[T: GenCodec](name: String)(values: Seq[T]): Unit = {
+    test(name) {
+      val serialized = values.map(write[T])
+      val deserialized = serialized.map(read[T])
+      deserialized shouldBe values
+    }
+  }
+
+  roundtrip("integers")(Seq(Int.MinValue, -42, 0, 42, Int.MaxValue))
+
+  roundtrip("longs")(Seq(Int.MaxValue.toLong + 1, Long.MinValue, -783868188, 0, 783868188, Long.MaxValue))
+
+  roundtrip("doubles")(Seq(
+    Double.MinPositiveValue, Double.MinValue, -1.750470182E9, Double.MaxValue, Double.PositiveInfinity, Double.NegativeInfinity)
+  )
+  roundtrip("booleans")(Seq(false, true))
+
+  roundtrip("strings")(Seq("", "a።bc\u0676ąቢść➔Ĳ"))
+
+  roundtrip("simple case classes")(Seq(TestCC(5, 123L, 432, true, "bla", 'a' :: 'b' :: Nil)))
+
+  test("NaN") {
+    val value = Double.NaN
+
+    val serialized = write(value)
+    val deserialized = read[Double](serialized)
+
+    assert(java.lang.Double.isNaN(deserialized))
+  }
+
+  test("scientific") {
+    val value = -1.750470182E9
+    val test = value.toString
+    val serialized = Seq("-1.750470182E+9", "-1.750470182E9", test)
+    val deserialized = serialized.map(read[Double])
+
+    deserialized should contain only value
```

Comment:
<rant>
Attempts to make Scala or any other language look like English sentences do not amuse me.
I'm a simple man, I just `assert`
</rant>

Reply:
<rant>
I don't care for your personal preferences, but to be perfectly honest I just want to have a beatiful error message when it fails. I can insert the dots if you like them :nerd_face: 
</rant>